### ISSUE TITLE
Connection errors when opening many parallel sessions

### ISSUE THREAD

**Issue by qix** [ronf/asyncssh#55]
Occasionally when opening many sessions at the same time over a single connection `asyncssh 1.5.3` seems to hit some race condition. The python backtrace doesn't look particularly interesting but is available at https://gist.github.com/qix/67a6d94f5307c8699afcbcd61c50c463 if it might help.

What did hint towards the race condition is the following output in `auth.log`:

```
May 27 18:46:30 sensu1 sshd[719]: error: buffer_get_ret: trying to get more bytes 19 than in buffer 14
May 27 18:46:30 sensu1 sshd[719]: error: buffer_get_string_ret: buffer_get failed
May 27 18:46:30 sensu1 sshd[719]: fatal: buffer_get_string: buffer error
May 27 18:46:30 sensu1 sshd[719]: pam_unix(sshd:session): session closed for user josh
May 27 18:46:30 sensu1 sshd[720]: error: buffer_get_ret: trying to get more bytes 20 than in buffer 15
May 27 18:46:30 sensu1 sshd[720]: error: buffer_get_string_ret: buffer_get failed
May 27 18:46:30 sensu1 sshd[720]: fatal: buffer_get_string: buffer error
May 27 18:46:30 sensu1 sshd[720]: pam_unix(sshd:session): session closed for user josh
May 27 18:46:30 sensu1 sshd[831]: fatal: mm_request_receive: read: Connection reset by peer
May 27 18:46:30 sensu1 sshd[832]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[830]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[828]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[834]: fatal: mm_request_receive: read: Connection reset by peer
May 27 18:46:30 sensu1 sshd[835]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[836]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[833]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[837]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[838]: fatal: mm_request_send: write: Broken pipe
May 27 18:46:30 sensu1 sshd[829]: fatal: mm_request_send: write: Broken pipe
```

**Comment by ronf** [comment]
Thanks for the report.

Unfortunately, as you said, the backtrace doesn't help much. Are you able to reproduce this well enough to track down what the specific SSH messages are that end up being the wrong length? Alternately, do you have a test case you could share so I could try to reproduce it here and add debugging around that?

**Comment by selfuryon** [comment]
I have the same issue. I develop asynchronous library for connecting to different network devices (Cisco, HP, Fujitsu and so on). And when i try to make more than 20-25 parallel sessions I have the problems... I tried to test parallel session by this simple script:
```python
import asyncio, asyncssh, sys

async def run_client():
    async with asyncssh.connect('localhost', username='self', password='***') as conn:
        result = await conn.run('pwd', check=True)
        print(result.stdout, end='')


async def run_clients():
    task = []
    for i in range (1,10):
        await run_client()
    print('end')


asyncio.get_event_loop().run_until_complete(run_clients())
```
When i use range with upper limit 10 - it's fine. But if i try to use 20 or more as upper limit I have this:
```
self:asyncssh-test/ $ python3 test1                                  [20:48:07]
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
/home/self
Task exception was never retrieved
future: <Task finished coro=<run_client() done, defined at test1:3> exception=DisconnectError('Disconnect Error: Connection lost',)>
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test1", line 4, in run_client
    async with asyncssh.connect('localhost', username='self', password='F0rReal!!') as conn:
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/misc.py", line 138, in __aenter__
    self._result = yield from self._coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4429, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4219, in create_connection
    yield from auth_waiter
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
asyncssh.misc.DisconnectError: Disconnect Error: Connection lost
Task exception was never retrieved
future: <Task finished coro=<run_client() done, defined at test1:3> exception=DisconnectError('Disconnect Error: Connection lost',)>
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test1", line 4, in run_client
    async with asyncssh.connect('localhost', username='self', password='F0rReal!!') as conn:
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/misc.py", line 138, in __aenter__
    self._result = yield from self._coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4429, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4219, in create_connection
    yield from auth_waiter
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
asyncssh.misc.DisconnectError: Disconnect Error: Connection lost
Task exception was never retrieved
future: <Task finished coro=<run_client() done, defined at test1:3> exception=DisconnectError('Disconnect Error: Connection lost',)>
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test1", line 4, in run_client
    async with asyncssh.connect('localhost', username='self', password='F0rReal!!') as conn:
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/misc.py", line 138, in __aenter__
    self._result = yield from self._coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4429, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4219, in create_connection
    yield from auth_waiter
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
asyncssh.misc.DisconnectError: Disconnect Error: Connection lost
Task exception was never retrieved
future: <Task finished coro=<run_client() done, defined at test1:3> exception=DisconnectError('Disconnect Error: Connection lost',)>
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "test1", line 4, in run_client
    async with asyncssh.connect('localhost', username='self', password='F0rReal!!') as conn:
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/misc.py", line 138, in __aenter__
    self._result = yield from self._coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4429, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 4219, in create_connection
    yield from auth_waiter
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
asyncssh.misc.DisconnectError: Disconnect Error: Connection lost
end

self:asyncssh-test/ $ python3 -V                                     
Python 3.5.2
self:asyncssh-test/ $ pip3 show asyncssh | grep Version            
Metadata-Version: 1.1
Version: 1.6.2

```

**Comment by ronf** [comment]
I'm not able to reproduce this here. Even setting it to 50 connections works fine. I'm testing against OpenSSH 7.2p1 here as my SSH server. Are you seeing the same buffer errors in the sshd log?

**Comment by selfuryon** [comment]
Hello! No, i don't see. I tried to run my test (https://gist.github.com/selfuryon/57cc5344c1130c643f2afbfc81600875) and this i got in auth.log with sshd loglevel verbose https://gist.github.com/selfuryon/384dded45186fcc4251bb1f0bb8d563d

**Comment by selfuryon** [comment]
I tried to run my test with PYTHONASYNCIODEBUG=1 and logging.basicConfig(level='DEBUG'). Output is here - https://gist.github.com/selfuryon/8f5e565a1d5fa979e77053a54eb327bf

**Comment by ronf** [comment]
I took a closer look at your sshd logs, and noticed that there are a series of "drop connection" messages there. I think the issue here might be code in sshd which is trying to protect against a denial-of-service attack, limiting the number of simultaneous connections going through authentication at once. That defaults to kicking in when you have 10 connections open:

```
     MaxStartups
             Specifies the maximum number of concurrent unauthenticated con-
             nections to the SSH daemon.  Additional connections will be
             dropped until authentication succeeds or the LoginGraceTime
             expires for a connection.  The default is 10:30:100.

             Alternatively, random early drop can be enabled by specifying the
             three colon separated values "start:rate:full'' (e.g.
             "10:30:60").  sshd(8) will refuse connection attempts with a
             probability of "rate/100'' (30%) if there are currently
             "start'' (10) unauthenticated connections.  The probability
             increases linearly and all connection attempts are refused if the
             number of unauthenticated connections reaches "full'' (60).
```

You can try raising this number and seeing if the problem goes away.

What's odd here, though, is that it looks like your test program should be starting one connection at a time, and waiting for it to complete before starting the next connection. So, I don't know why you'd be getting so many connections in parallel. The log file you pasted suggests you are getting lots of connections at once, though. Are you really doing an "await" on run_client() as shown above, or are you creating multiple async tasks at once before waiting for them to complete?

**Comment by selfuryon** [comment]
Yes, sorry. It was my first atempt for making test. Full code that i run was 
```python
import asyncio, asyncssh, sys


async def run_client():
    async with asyncssh.connect('localhost', username='self', password='***') as conn:
        result = await conn.run('pwd', check=True)
        print(result.stdout, end='')


async def run_clients():
    task = []
    for i in range (1,20):
        task.append(run_client())
    await asyncio.wait(task)
    print('end')


asyncio.get_event_loop().run_until_complete(run_clients())
```
And really after increasing MaxStartups I able to run many parallel sessions. Thanks a lot!!

**Comment by ronf** [comment]
Yeah, ok - the new code there would be starting all the sessions in parallel, so it makes sense that MaxStartups would be an issue. Glad that worked for you!

While this doesn't resolve the original issue qix posted here, I'm going to close this for now since I never got more details on that and was unable to reproduce it here. If anyone sees this problem again, please feel free to open a new issue!
